quodlibet/util/collection.py

```diff
@@ -47,19 +48,73 @@
 ELPOEP = list(reversed(PEOPLE))
 PEOPLE_SCORE = [100 ** i for i in range(len(PEOPLE))]
 
+SMOOTHING_FACTOR_DEFAULT = 10
 
-def avg(nums):
+
+def unweighted_average(nums):
     """Returns the average (arithmetic mean) of a list of numbers"""
     return float(sum(nums)) / len(nums)
 
 
-def bayesian_average(nums, c=None, m=None):
+def unweighted_bayesian_average(nums, c=None, m=None):
     """Returns the Bayesian average of an iterable of numbers,
     with parameters defaulting to config specific to ~#rating."""
     m = m or config.RATINGS.default
     c = c or config.getfloat("settings", "bayesian_rating_factor", 0.0)
-    ret = float(m * c + sum(nums)) / (c + len(nums))
-    return ret
+    return float(m * c + sum(nums)) / (c + len(nums))
+
+
+def hyperbolic_smoother(x, n=SMOOTHING_FACTOR_DEFAULT):
+    """n determines how strongly smaller values will be magnified"""
+    return (n + 1) * x / (n * x + 1)
+
+
+def smoothed_length_weights_and_nums(songs, key, total_length, smoothing_factor):
+    """Returns smoothed length-based weights and the songs' associated numeric values"""
+    length_based_smoothed_weights = []
+    values = []
+    for s in songs:
+        value = s(key)
+        if value == "":
+            continue
+
+        length = s("~#length", 0)
+        if length <= 0:
+            # will have no impact on the average, so we can just skip it
+            continue
+
+        values.append(value)
+        weight = hyperbolic_smoother(length / total_length, smoothing_factor)
+
+        length_based_smoothed_weights.append(weight)
+
+    return length_based_smoothed_weights, values
+
+
+def weighted_average(nums, weights):
+    """Returns the weighted average (arithmetic mean) of a list of numbers"""
+    total_weight = 0
+    weighted_sum = 0
+    for num, weight in zip(nums, weights):
+        total_weight += weight
+        weighted_sum += weight * num
+
+    return weighted_sum / total_weight
+
+
+def weighted_bayesian_average(nums, weights, c=None, m=None):
+    """Returns the weighted Bayesian average of an iterable of numbers,
+    with parameters defaulting to config specific to ~#rating."""
+    m = m or config.RATINGS.default
```

Comment:
Probably being dim but I don't get why the default rating has an influence here?

Reply:
This is copied from the previous unweighted Bayesian average. I think, after reading the Wiki article again, that the default rating is chosen, because it seems like a reasonable prior mean.. 

Simplified, but the following also helped me: https://old.reddit.com/r/explainlikeimfive/comments/bls5hq/eli5_what_is_a_bayesian_average_used_for_and_how/emr2mqj/ :smile: 

> Calculating the Bayesian average uses the prior mean m and a constant C.

@declension Thank you for the review! :smiley:

Reply:
lol so I'm asking questions about my own code, classic

Reply:
I know that feeling.  Sometimes reading code one wrote in the past is really like reading somebody else's code. And in some sense that's true, as we're changing and hopefully improving too.